Objective-C/Internal/Replicator/CBLWebSocket.mm

```diff
@@ -372,63 +370,79 @@ - (void) connectToHostWithName: (NSString*)hostname
         CBLWarnError(WebSocket, @"%@: %@", self, msg);
-        [self closeWithError: posixError(EPERM, msg)];
+        [self closeWithError: posixError(errNo, msg)];
         return;
     }
-    
-    // Create socket:
-    [self _socketConnect: interface];
+
+    _dnsService = [[CBLDNSService alloc] initWithHost: hostname
+                                            interface: index
+                                                 port: (UInt16)port
+                                             delegate: self];
+    [_dnsService start];
 }
 
+#pragma mark DNSServiceDelegate
+
+ - (void) didResolveSuccessWithAddress: (AddressInfo*)info {
+     dispatch_async(_queue, ^{
+         if (_dnsService) {
+             CBLLogVerbose(WebSocket, @"%@: Host '%@' was resolved as ip=%@, family=%d",
+                           self, info.host, info.addrstr, info.addr->sa_family);
+             [self _socketConnect: info];
+         }
+     });
+ }
+
+ - (void) didResolveFailWithError: (NSError*)error {
+     dispatch_async(_queue, ^{
+         if (_dnsService) {
+             CBLWarnError(WebSocket, @"%@: Host was failed to resolve with error '%@'",
+                           self, error.my_compactDescription);
+             [self closeWithError: error];
+         }
+     });
+ }
+
 #pragma mark - Socket connect
 
-- (void) _socketConnect: (NSString*)interface {
+- (void) _socketConnect: (AddressInfo*)info {
     Assert(_sockfd < 0);
-    _sockfd = socket(_addr->ai_family, _addr->ai_socktype, _addr->ai_protocol);
+    _sockfd = socket(info.addr->sa_family, SOCK_STREAM, 0);
     if (_sockfd < 0) {
         int errNo = errno;
-        NSString* msg = $sprintf(@"Failed to create socket with errno %d (%@)", errNo, addrInfo(_addr));
+        NSString* msg = $sprintf(@"Failed to create socket with errno %d (%@)", errNo, info);
         CBLWarnError(WebSocket, @"%@: %@", self, msg);
         [self closeWithError: posixError(errNo, msg)];
         return;
     }
     
     // Set network interface:
-    if (interface) {
-        NSError* err;
-        bool useIPv4 = (_addr->ai_family == AF_INET);
+    if (_networkInterface) {
+        UInt32 index = info.interface;
+        CBLLogVerbose(WebSocket, @"%@: Interface '%@' is mapped to index '%u'", self, _networkInterface, index);
+        int result = -1;
+        if (info.addr->sa_family == AF_INET) {
+            result = setsockopt(_sockfd, IPPROTO_IP, IP_BOUND_IF, &index, sizeof(index));
+        } else if (info.addr->sa_family == AF_INET6 ){
+            result = setsockopt(_sockfd, IPPROTO_IPV6, IPV6_BOUND_IF, &index, sizeof(index));
+        }
         
-        if (![self setSocketOptForInterface: interface useIPv4: useIPv4 error: &err]) {
-            [self closeWithError: err];
+        if (result < 0) {
+            int errNo = errno;
+            NSString* msg = $sprintf(@"Failed to set network interface %@ with errno %d (%@)",
+                                     _networkInterface, errNo, info);
```

Comment:
As there is already a log about mapping network interface to index above, here we can log with just index as this method only uses index.